torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp

```diff
@@ -599,494 +641,101 @@ TORCH_CUDA_CU_API c10::optional<ReductionParams> getReductionHeuristics(
   }
 
   size_t max_dtype_size = 1;
-  size_t n_input_tensors = 0;
+  size_t n_tensor_inputs = 0;
   for (auto inp : fusion->inputs()) {
     if (inp->isA<TensorView>()) {
       max_dtype_size =
           std::max(max_dtype_size, dataTypeSize(inp->getDataType().value()));
-      n_input_tensors++;
+      n_tensor_inputs++;
     }
   }
 
   TORCH_INTERNAL_ASSERT(
-      n_input_tensors > 0,
+      n_tensor_inputs > 0,
       "Tried to schedule a fusion with no tensor inputs, currently not supported.");
 
+  auto vectorizable_inputs_outputs =
+      scheduler_utils::getVectorizableInputsOutputs(reduction_tv);
+
+  // Vectorize as much as we can
+  size_t vectorize_factor = std::numeric_limits<size_t>::max();
+
+  for (auto tv : vectorizable_inputs_outputs) {
+    const auto tv_vectorize_factor = runtime_info.getVectorizableWidth(tv);
+    vectorize_factor = std::min(vectorize_factor, tv_vectorize_factor);
+  }
+
+  if (vectorize_factor == std::numeric_limits<size_t>::max()) {
+    vectorize_factor = 1;
+  }
+
   return reductionHeuristic(
       red_elements,
       num_outputs_for_reduction,
       fastest_dim_reduction,
-      n_input_tensors,
-      max_dtype_size);
+      n_tensor_inputs,
+      max_dtype_size,
+      vectorize_factor);
 }
 
 // fusion is the input IR that will be modified by this function
 void scheduleReduction(Fusion* fusion, const ReductionParams& rparams) {
```

Comment:
Majority of logic has been moved to scheduler utils, it's now shared across reduction, multi-reduction, and normalization.